### PR TITLE
Add language extensions counters

### PR DESCRIPTION
**Pull request by milan-tom**
Adds some basic counters for the usages of the following language extensions:

- `comprehensions`
- `include functor`
- `immutable arrays`
- `module strengthening`
- `labeled tuples`
- `immutable arrays`

### CODE REVIEW

File: ocaml/driver/compile_common.ml [hunk 1]
```diff
@@ -123,7 +129,13 @@ let parse_impl i =
...
 let typecheck_impl i parsetree =
   parsetree
-  |> Profile.(record typing)
+  |> Profile.(
+    record_with_counters
+      ~counter_f:(fun (typed_tree : Typedtree.implementation) ->
+        Profile_counters_functions.(
+          count_language_extensions
+            (Typedtree_implementation_output typed_tree)))
+      typing)
```
Comment (xclerc) [inline ocaml/driver/compile_common.ml:138]: (Maybe factor out with lines 63-69,
if possible / easy.)

File: ocaml/utils/profile_counters_functions.ml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            | Texp_list_comprehension _ -> incr Comprehensions
+            | Texp_array_comprehension (mutability, _, _) ->
+              incr Comprehensions;
+              check_array_mutability mutability
+            | Texp_array (mutability, _, _, _) ->
+              check_array_mutability mutability
+            | Texp_tuple (label_opt_pair_list, _) ->
+              check_for_labeled_tuples label_opt_pair_list
+            | _ -> ());
+            default_iterator.expr sub e);
```
Comment (xclerc) [inline ocaml/utils/profile_counters_functions.ml:79]: I am not sure it really matters, but does
it count properly if there are _e.g._ nested
comprehensions or a labelled tuple inside
a comprehension?

**Review by xclerc — APPROVED** [review]
As discussed off-line, a follow-up may be
useful to ensure that the reference files
are stable if/when other counters are added.

File: ocaml/testsuite/tests/profile/counters/language_extensions/labeled_tuples/ml_counters.ocamlc.reference [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ labeled_tuples/ml_counters.ml
+  [comprehensions = 0; immutable_arrays = 0; include_functor = 0; labeled_tuples = 4; module_strengthening = 0] typing
```
[the remark this answers is not in the document]
Comment (milan-tom) [reply ocaml/testsuite/tests/profile/counters/language_extensions/labeled_tuples/ml_counters.ocamlc.reference:2]: The comment in the other file was incorrect. I have corrected it.

File: ocaml/testsuite/tests/profile/counters/language_extensions/labeled_tuples/ml_counters.ml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+(* Labeled tuples usage *)
+
+(* 1 type, 1 expression*)
```
[the remark this answers is not in the document]
Comment (milan-tom) [reply ocaml/testsuite/tests/profile/counters/language_extensions/labeled_tuples/ml_counters.ml]: Yes, that was incorrect. I have changed it to read `1 expression`.

File: ocaml/driver/compile_common.ml (hunk 1 above)
Comment (milan-tom) [reply ocaml/driver/compile_common.ml:138]: It seems slightly less readable factored out as the two use different variant constructors (not treated as functions in OCaml) and one uses `Profile.record_with_counters` whereas the other uses `Profile.record_call_with_counters`.

File: ocaml/utils/profile_counters_functions.ml (hunk 2 above)
Comment (milan-tom) [reply ocaml/utils/profile_counters_functions.ml:79]: It should as the iterator uses open recursion over the type tree. I will add some tests.

Comment (milan-tom) [reply ocaml/utils/profile_counters_functions.ml:79]: Added

**Pull request merged**